internal/fxlog/fxlog.go

```diff
@@ -21,68 +21,117 @@
 package fxlog
 
 import (
-	"fmt"
-	"log"
-	"os"
-	"strings"
-
-	"go.uber.org/fx/internal/fxreflect"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
-var _exit = func() { os.Exit(1) }
+// Level is the level of logging used by Logger.
+type Level int
+
+const (
+	// InfoLevel is Info logging level used to log messages via zap.
+	InfoLevel Level = iota
+	// ErrorLevel is Info logging level used to log messages via zap.
+	ErrorLevel
+)
 
-// Printer is a formatting printer.
-type Printer interface {
-	Printf(string, ...interface{})
+// Entry is an entry to be later serialized into zap message and fields.
+type Entry struct {
+	Level   Level
+	Message string
+	Fields  []Field
+	Stack   string
 }
 
-// New returns a new Logger backed by the standard library's log package.
-func New() *Logger {
-	return &Logger{log.New(os.Stderr, "", log.LstdFlags)}
+// WithStack mutates an Entry to add a Stack field.
+func (e Entry) WithStack(stack string) Entry {
+	e.Stack = stack
+	return e
 }
 
-// A Logger writes output to standard error.
-type Logger struct {
-	Printer
+func (e Entry) Write(logger Logger) {
+	logger.Log(e)
 }
 
-// Printf logs a formatted Fx line.
-func (l *Logger) Printf(format string, v ...interface{}) {
-	l.Printer.Printf(prepend(format), v...)
+// Err is a helper for error Fields.
+func Err(value interface{}) Field {
+	return F("error", value)
 }
 
-// PrintProvide logs a type provided into the dig.Container.
-func (l *Logger) PrintProvide(t interface{}) {
-	for _, rtype := range fxreflect.ReturnTypes(t) {
-		l.Printf("PROVIDE\t%s <= %s", rtype, fxreflect.FuncName(t))
+// F is a constructor for a Field.
+func F(key string, value interface{}) Field {
+	return Field{
+		Key:   key,
+		Value: value,
 	}
 }
 
-// PrintSupply logs a type supplied directly into the dig.Container
-// by the given constructor function.
-func (l *Logger) PrintSupply(constructor interface{}) {
-	for _, rtype := range fxreflect.ReturnTypes(constructor) {
-		l.Printf("SUPPLY\t%s", rtype)
+// Field defines a field used inside an internal logging Entry.
+type Field struct {
+	Key   string
+	Value interface{}
+}
+
+// Logger defines interface used for logging.
+type Logger interface {
+	Log(entry Entry)
+}
+
+var _ Logger = (*zapLogger)(nil)
+
+type zapLogger struct {
+	logger *zap.Logger
+}
+
+func encodeFields(fields []Field, stack string) []zap.Field {
+	var fs []zap.Field
+	for _, field := range fields {
+		fs = append(fs, zap.Any(field.Key, field.Value))
+	}
+	if stack != "" {
+		fs = append(fs, zap.Stack(stack))
```

Comment:
this doesn't work like we want it to work.

https://pkg.go.dev/go.uber.org/zap#Stack accepts the name of the JSON key that should contain the stack trace, but it captures its own stack trace.

For now, I would do this:

```suggestion
	if stack != "" {
		fs = append(fs, zap.String("stack", stack))
```